backends/ebpf/psa/externs/ebpfPsaTableImplementation.cpp

```diff
@@ -206,4 +207,446 @@ void EBPFActionProfilePSA::applyImplementation(CodeBuilder* builder, cstring tab
+                                   cstring("struct ") + valueTypeName, 1);
+
+    // action map (ref -> action)
+    builder->target->emitTableDecl(builder, actionsMapName, TableHash, "u32",
+                                   cstring("struct ") + valueTypeName, size);
+
+    // TODO: cache instance
+}
+
+void EBPFActionSelectorPSA::emitReferenceEntry(CodeBuilder *builder) {
+    EBPFTableImplementationPSA::emitReferenceEntry(builder);
+    builder->emitIndent();
+    builder->appendFormat("u32 %s", isGroupEntryName.c_str());
+    builder->endOfStatement(true);
+}
+
+void EBPFActionSelectorPSA::applyImplementation(CodeBuilder* builder, cstring tableValueName,
+                                                cstring actionRunVariable) {
+    cstring msg = Util::printf_format("ActionSelector: applying %s", instanceName.c_str());
+    builder->target->emitTraceMessage(builder, msg.c_str());
+
+    // 1. Declare variables.
+
+    cstring asValueName = program->refMap->newName("as_value");
+    cstring effectiveActionRefName = program->refMap->newName("as_action_ref");
+    cstring innerGroupName = program->refMap->newName("as_group_map");
+    groupStateVarName = program->refMap->newName("as_group_state");
+    // these can be hardcoded because they are declared inside of a block
+    cstring checksumValName = "as_checksum_val";
+    cstring mapEntryName = "as_map_entry";
+
+    // TODO: cache variables
+
+    builder->emitIndent();
+    builder->appendFormat("struct %s * %s = NULL", valueTypeName.c_str(), asValueName.c_str());
+    builder->endOfStatement(true);
+    builder->emitIndent();
+    builder->appendFormat("u32 %s = %s->%s", effectiveActionRefName.c_str(),
+                          tableValueName.c_str(), referenceName.c_str());
+    builder->endOfStatement(true);
+    builder->emitIndent();
+    builder->appendFormat("u8 %s = 0", groupStateVarName.c_str());
+    builder->endOfStatement(true);
+
+    // 2. Check if we have got group reference.
+
+    builder->emitIndent();
+    builder->appendFormat("if (%s->%s != 0) ", tableValueName.c_str(), isGroupEntryName.c_str());
+    builder->blockStart();
+    builder->target->emitTraceMessage(builder, "ActionSelector: group reference %u",
+                                      1, effectiveActionRefName.c_str());
+
+    // TODO: cache lookup
+
+    builder->emitIndent();
+    builder->append("void * ");
+    builder->target->emitTableLookup(builder, groupsMapName, effectiveActionRefName,
+                                     innerGroupName);
+    builder->endOfStatement(true);
+    builder->emitIndent();
+    builder->appendFormat("if (%s != NULL) ", innerGroupName.c_str());
+    builder->blockStart();
+
+    // 3. Find member reference.
+    // 3.1. First entry in inner map contains number of valid elements in the map
+
+    builder->emitIndent();
+    builder->appendFormat("u32 * %s = bpf_map_lookup_elem(%s, &%s)", mapEntryName.c_str(),
```

Comment:
can't we use `emitTableLookup()` here?

Reply:
Because it assumes that map is not a pointer, but `innerGroupName` is a pointer to inner map. `emitTableLookup()` emits:
```c
builder->appendFormat("BPF_MAP_LOOKUP_ELEM(%s, &%s)", tblName.c_str(), key.c_str());
```
Where macro is defined as follow:
```c
#define BPF_MAP_LOOKUP_ELEM(table, key) bpf_map_lookup_elem(&table, key)
```